src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,265 @@
+                    auto user_clone_ptr = typeid_cast<std::shared_ptr<User>>(clone);
+                    user_clone_ptr->granted_roles.grant(id);
+                    return user_clone_ptr;
+                }
+                return cached_entity;
+            };
+
+            memory_storage.update(memory_storage.findAll<User>(), update_func);
+            roles_of_interest.insert(id);
+        }
+    }
+    else
+    {
+        if (roles_of_interest.find(id) != roles_of_interest.end())
+        {
+            auto update_func = [&id](const AccessEntityPtr & cached_entity) -> AccessEntityPtr
+            {
+                auto user_ptr = typeid_cast<std::shared_ptr<const User>>(cached_entity);
+                if (user_ptr && user_ptr->granted_roles.roles.find(id) != user_ptr->granted_roles.roles.end())
+                {
+                    auto clone = user_ptr->clone();
+                    auto user_clone_ptr = typeid_cast<std::shared_ptr<User>>(clone);
+                    user_clone_ptr->granted_roles.revoke(id);
+                    return user_clone_ptr;
+                }
+                return cached_entity;
+            };
+
+            memory_storage.update(memory_storage.findAll<User>(), update_func);
+            roles_of_interest.erase(id);
+        }
+    }
+}
+
+
+const char * LDAPAccessStorage::getStorageType() const
+{
+    return STORAGE_TYPE;
+}
+
+
+bool LDAPAccessStorage::isStorageReadOnly() const
+{
+    return true;
+}
+
+
+std::optional<UUID> LDAPAccessStorage::findImpl(EntityType type, const String & name) const
+{
+    return memory_storage.find(type, name);
+}
+
+
+std::optional<UUID> LDAPAccessStorage::findOrGenerateImpl(EntityType type, const String & name) const
+{
+    if (type == EntityType::USER)
+    {
+        std::scoped_lock lock(mutex);
+
+        // Return the id immediately if we already have it.
+        const auto id = memory_storage.find(type, name);
+        if (id.has_value())
+            return id;
+
+        if (!isConfiguredNoLock())
+            return {};
+
+        // Stop if entity exists anywhere else, to avoid generating duplicates.
+        const auto * this_base = dynamic_cast<const IAccessStorage *>(this);
+        const auto storages = access_control_manager->getStoragesPtr();
+        for (const auto & storage : *storages)
+        {
+            if (storage.get() != this_base && storage->find(type, name))
+                return {};
+        }
+
+        // Entity doesn't exist. We are going to create one.
+        const auto user = std::make_shared<User>();
+        user->setName(name);
+        user->authentication = Authentication(Authentication::Type::LDAP_SERVER);
+        user->authentication.setServerName(ldap_server);
+
+        for (const auto& role_name : default_role_names)
+        {
+            std::optional<UUID> role_id;
+
+            try
+            {
+                role_id = access_control_manager->find<Role>(role_name);
```

Comment:
You don't have to search for roles all the time. Find them just once in the constructor or when the first user is created, then subscribe for changes, and just assign `user->granted_roles = roles_of_intereset` here.

Reply:
If any of roles doesn't exist at the time I try to authenticate a user, I need to reject the attempt (at least this was the formal behavior we fixed.) Later, for cached User instances, roles may be removed and re-added, and I will track and apply these changes accordingly.

Reply:
>If any of roles doesn't exist at the time I try to authenticate a user, I need to reject the attempt

That seems kind of overkill, but ok.

However you have a subscription for any changes of all the roles, so if any role you need is added anytime you will know about that immediately because `LDAPAccessStorage::processRoleChange()` will be called. Thus you can simply track roles which you need but haven't get yet and this way you don't need to find roles in `AccessControlManager`.

Reply:
You mean that I can track role `id`s to avoid lookups when I need to authenticate a newly created User in `LDAPAccessStorage::loginImpl()`?

Reply:
You're already tracking them because you subscribe for changing of roles. I suggest just to use the result of that tracking.
And it's definitely not necessary to find exactly the same roles when you create every user.

Reply:
What if roles I am interested in are added before I am subscribed for changes?

Reply:
> I suggest just to use the result of that tracking. And it's definitely not necessary to find exactly the same roles when you create every user.

Something like the following code should work (see also https://github.com/ClickHouse/ClickHouse/pull/12736#discussion_r500566699):

```
UUID LDAPAccessStorage::loginImpl(const String & user_name, const String & password, const Poco::Net::IPAddress & address, const ExternalAuthenticators & external_authenticators) const
{
        ...
        if (!isAddressAllowedImpl(*user, address))
            throwAddressNotAllowed(address);

        checkAllDefaultRoleNamesFound();
        auto & granted_roles = user->granted_roles.roles;
        boost::range::copy(roles_of_interest | boost::adaptors::map_keys, std::inserter(granted_roles));

        return memory_storage.insert(user);
    }
}

void LDAPAccessStorage::checkAllDefaultRoleNamesFound()
{
    if (default_role_names.size() == roles_of_interest.size())
        return;

    boost::container::flat_set<std::string_view> not_found_role_names;
    boost::range::copy(default_role_names, std::inserter(not_found_role_names));
    for (const auto & role_name : roles_of_interest | boost::adaptors::map_values)
        not_found_role_names.erase(role_name);

    if (not_found_role_names.empty())
        return;

    throwDefaultRoleNotFound(*not_found_role_names.begin());
}
```

Reply:
Pushed the proposed changes with slight modifications.